autoload/ale/rename.vim

```diff
@@ -0,0 +1,193 @@
+            let l:text_changes = []
+
+            for l:edit in l:text_edits
+                let l:range = l:edit.range
+                let l:new_text = l:edit.newText
+
+                call add(l:text_changes, {
+                \ 'start': {
+                \   'line': l:range.start.line + 1,
+                \   'offset': l:range.start.character + 1,
+                \ },
+                \ 'end': {
+                \   'line': l:range.end.line + 1,
+                \   'offset': l:range.end.character + 1,
+                \ },
+                \ 'newText': l:new_text,
+                \})
+            endfor
+
+            call add(l:changes, {
+            \   'fileName': l:file_name,
+            \   'textChanges': l:text_changes,
+            \})
+        endfor
+
+        if empty(l:changes)
+            call ale#util#Execute('echom ''Could not rename.''')
+        else
+            call ale#code_action#HandleCodeAction({
+            \   'description': 'rename',
+            \   'changes': l:changes,
+            \})
+        endif
+    endif
+endfunction
+
+function! s:OnReady(line, column, new_name, linter, lsp_details) abort
+    let l:id = a:lsp_details.connection_id
+
+    if !ale#lsp#HasCapability(l:id, 'rename')
+        return
+    endif
+
+    let l:buffer = a:lsp_details.buffer
+
+    let l:Callback = a:linter.lsp is# 'tsserver'
+    \   ? function('ale#rename#HandleTSServerResponse')
+    \   : function('ale#rename#HandleLSPResponse')
+
+    call ale#lsp#RegisterCallback(l:id, l:Callback)
+
+    if a:linter.lsp is# 'tsserver'
+        let l:message = ale#lsp#tsserver_message#Rename(
+        \   l:buffer,
+        \   a:line,
+        \   a:column
+        \)
+    else
+        " Send a message saying the buffer has changed first, or the
+        " rename position probably won't make sense.
+        call ale#lsp#NotifyForChanges(l:id, l:buffer)
+
+        let l:message = ale#lsp#message#Rename(
+        \   l:buffer,
+        \   a:line,
+        \   a:column,
+        \   a:new_name
+        \)
+    endif
+
+    let l:request_id = ale#lsp#Send(l:id, l:message)
+
+    let s:rename_map[l:request_id] = {
+    \   'new_name': a:new_name,
+    \}
+endfunction
+
+function! s:ExecuteRename(linter, new_name) abort
+    let l:buffer = bufnr('')
+    let [l:line, l:column] = getpos('.')[1:2]
+
+    if a:linter.lsp isnot# 'tsserver'
+        let l:column = min([l:column, len(getline(l:line))])
+    endif
+
+    let l:Callback = function('s:OnReady', [l:line, l:column, a:new_name])
+    call ale#lsp_linter#StartLSP(l:buffer, a:linter, l:Callback)
+endfunction
+
+function! ale#rename#Execute(...) abort
+    let l:lsp_linters = []
+
+    for l:linter in ale#linter#Get(&filetype)
+        if !empty(l:linter.lsp)
+            call add(l:lsp_linters, l:linter)
+        endif
+    endfor
+
+    if !empty(l:lsp_linters)
+        let l:prompt = 'new name: '
+        let l:new_name = a:0 ? join(a:000) : input(l:prompt, expand('<cWORD>'))
```

Comment:
Instead of supporting an optional argument for a new name in the command, always require the new name to be input with `input()` and the current name. Then you always know exactly what name ALE is going to try and rename, and it's easy to adjust the current name to something slightly different. See here: https://github.com/dense-analysis/ale/commit/ca2026bd13346cb7e1afeecb54dbec3da0bd5151#diff-c8e42b11378165eaf5fb1be0e53860e6R90

Reply:
I implemented this change but since `feedkeys()` doesn't work for input prompt, I added a new util method `ale#util#Input(message, value)` and mocked that method in the test. Let me know if this works!